vtex/faststore
# feat: Add skuSelector-related properties to ProductGroup type

## What's the purpose of this pull request?

This PR adds a new attribute (`skuVariations`) to the `ProductGroup` type, that brings with it 4 new properties:

- `slugsMap`
- `availableVariations`
- `allVariantsByName`
- `activeVariations`

They should enable users to query data about SKU specification variants in a more convenient way. Specially handy for implementing SKU Selector components.

## How it works?

The new fields added to `ProductGroup` are all related to SKU variants.

**slugsMap** -> Maps property value combinations to their respective SKU's slug. Enables us to retrieve the slug for the SKU that matches the currently selected variations in `O(1)` time.

**activeVariations** -> SKU property values for the current SKU.

**availableVariations** -> Available options for each varying SKU property, taking into account the `dominantVariantName` property. Returns all available options for the dominant property, and only options that can be combined with its current value for other properties.

**allVariantsByName** -> All available options for each SKU variant property, indexed by their name.

This is an example query that only asks for the new fields:

```graphql
{
  product(locator: {key: "id", value: "36"}) {
    isVariantOf {
      skuVariants {
        activeVariations
        allVariantsByName
        slugsMap(dominantVariantName: "Color")
        availableVariations(dominantVariantName: "Color")
      }
    }
  }
}
```

And this is how a response would look like:

```json
{
  "data": {
    "product": {
      "isVariantOf": {
        "skuVariants": {
          "activeVariations": {
            "Color": "White",
            "Size": "42"
          },
          "allVariantsByName": {
            "Color": [
              "Green",
              "White",
              "Red"
            ],
            "Size": [
              "40",
              "42",
              "41"
            ]
          },
          "slugsMap": {
            "Color-Red-Size-42": "classic-shoes-37",
            "Color-Green-Size-40": "classic-shoes-35",
            "Color-White-Size-42": "classic-shoes-36",
            "Color-White-Size-41": "classic-shoes-310124175"
          },
          "availableVariations": {
            "Color": [
              {
                "src": "https://storecomponents.vtexassets.com/arquivos/ids/155560/pink-cool-sku-variation.png?v=637063239809000000",
                "alt": "skuvariation",
                "label": "Red",
                "value": "Red"
              },
              {
                "src": "https://storecomponents.vtexassets.com/arquivos/ids/155559/pink-sku-variation.png?v=637087508159070000",
                "alt": "skuvariation",
                "label": "Green",
                "value": "Green"
              },
              {
                "src": "https://storecomponents.vtexassets.com/arquivos/ids/155561/white-sku-variation.png?v=637087507771770000",
                "alt": "skuvariation",
                "label": "White",
                "value": "White"
              }
            ],
            "Size": [
              {
                "src": "https://storecomponents.vtexassets.com/arquivos/ids/155561/white-sku-variation.png?v=637087507771770000",
                "alt": "skuvariation",
                "label": "42",
                "value": "42"
              },
              {
                "src": "https://storecomponents.vtexassets.com/arquivos/ids/155561/white-sku-variation.png?v=637087507771770000",
                "alt": "skuvariation",
                "label": "41",
                "value": "41"
              }
            ]
          }
        }
      }
    }
  }
}
```

## How to test it?

You can either run this branch locally by running `yarn develop` inside `/packages/api` or check these fields already being used to render an SKU Selector in [this PR](https://github.com/vtex-sites/nextjs.store/pull/158).

### Starters Deploy Preview

https://sfj-5b76285--nextjs.preview.vtex.app/classic-shoes/p. -> from https://github.com/vtex-sites/nextjs.store/pull/158
